emscripten-simulated-jerry-api/jerry.js

```diff
@@ -148,6 +151,81 @@
       this._jerryToHostValueMap = {};
       this._nextObjectRef = 1;
       this._jerryInternalPropsWeakMap = new WeakMap();
+
+      if (!Module.Promise) {
+        console.warn('Please set Module.Promise to a promise implementation to use for JerryScript')
+      } else {
+        // Patch up the Promise.prototype then/catch implementations.
+        // The patched versions should create a function that when called, pushes the original callback arg to the jobQueue
+        var _then = Module.Promise.prototype.then;
+        var _catch = Module.Promise.prototype.catch;
+        var _identify = function(x) {return x};
+        var _thrower = function(x) {throw x};
+        var jobQueue = this.jobQueue;
+
+        // Modify the `then`.
+        Module.Promise.prototype.then = function (cb1, cb2) {
+          var res, rej;
+          var ret = new Module.Promise(function(f, r) {
+            res = f;
+            rej = r;
+          });
+          if (typeof cb1 !== 'function') {
+            cb1 = _identify;
+          }
+          if (typeof cb2 !== 'function') {
+            cb2 = _thrower;
+          }
+
+          var newCb1 = function(arg) {
+            jobQueue.push({arg: arg, cb: cb1, resolve:res, reject:rej});
+            return ret;
+          };
+
+          var newCb2 = function(arg) {
+            jobQueue.push({arg: arg, cb: cb2, resolve:res, reject:rej});
+            return ret;
+          };
+
+          return _then.call(this, newCb1, newCb2);
+        };
+
+        // Modify the `catch`.
+        Module.Promise.prototype.catch = function (cb) {
+          var res, rej;
+          var ret = new Module.Promise(function(f, r) {
+            res = f;
+            rej = r;
+          });
+
+          if (typeof cb !== 'function') {
+            cb = _thrower;
+          }
+
+          var newCb = function(arg) {
+            jobQueue.push({arg: arg, cb: cb, resolve:res, reject:rej});
+            return ret;
+          };
+
+          return _catch.call(this, newCb);
+        };
+      }
+    },
+
+    runJobQueue: function() {
+      var job = this.jobQueue.pop();
```

Comment:
`pop` takes from the back of the queue, should do `shift()` instead?

Reply:
Oh, forget that